File: frontend/services/auth_service.py
"""
Auth Service
Handles authentication and authorization
"""
from utils.api_client import api_client
from typing import Dict, Optional
import streamlit as st


class AuthService:
    """Authentication service"""
    
    @staticmethod
    def login(username: str, password: str) -> Dict:
        """
        Login user
        Returns user data with access token
        """
        data = {
            "username": username,
            "password": password
        }
        response = api_client.post("/api/auth/login", data)
        
        # Store in session
        if response:
            st.session_state.authenticated = True
            st.session_state.access_token = response.get("access_token")
            st.session_state.refresh_token = response.get("refresh_token")
            st.session_state.user = response.get("user")
            st.session_state.user_role = response.get("user", {}).get("role")
        
        return response
    
    @staticmethod
    def logout():
        """Logout user"""
        try:
            api_client.post("/api/auth/logout")
        except:
            pass
        
        # Clear session
        for key in list(st.session_state.keys()):
            del st.session_state[key]
    
    @staticmethod
    def register(data: Dict) -> Dict:
        """Register new user"""
        return api_client.post("/api/auth/register", data)
    
    @staticmethod
    def refresh_token() -> Dict:
        """Refresh access token"""
        refresh_token = st.session_state.get("refresh_token")
        if not refresh_token:
            raise Exception("No refresh token available")
        
        data = {"refresh_token": refresh_token}
        response = api_client.post("/api/auth/refresh", data)
        
        if response:
            st.session_state.access_token = response.get("access_token")
        
        return response
    
    @staticmethod
    def change_password(current_password: str, new_password: str) -> Dict:
        """Change user password"""
        data = {
            "current_password": current_password,
            "new_password": new_password
        }
        return api_client.post("/api/auth/change-password", data)
    
    @staticmethod
    def request_password_reset(email: str) -> Dict:
        """Request password reset"""
        data = {"email": email}
        return api_client.post("/api/auth/password-reset/request", data)
    
    @staticmethod
    def reset_password(token: str, new_password: str) -> Dict:
        """Reset password with token"""
        data = {
            "token": token,
            "new_password": new_password
        }
        return api_client.post("/api/auth/password-reset/confirm", data)
    
    @staticmethod
    def verify_email(token: str) -> Dict:
        """Verify email with token"""
        data = {"token": token}
        return api_client.post("/api/auth/verify-email", data)
    
    @staticmethod
    def setup_2fa() -> Dict:
        """Setup two-factor authentication"""
        return api_client.post("/api/auth/2fa/setup")
    
    @staticmethod
    def verify_2fa(code: str) -> Dict:
        """Verify 2FA code"""
        data = {"code": code}
        return api_client.post("/api/auth/2fa/verify", data)
    
    @staticmethod
    def disable_2fa() -> Dict:
        """Disable two-factor authentication"""
        return api_client.post("/api/auth/2fa/disable")
    
    @staticmethod
    def is_authenticated() -> bool:
        """Check if user is authenticated"""
        return st.session_state.get("authenticated", False)
    
    @staticmethod
    def get_current_user() -> Optional[Dict]:
        """Get current logged in user"""
        return st.session_state.get("user")
    
    @staticmethod
    def get_user_role() -> Optional[str]:
        """Get current user role"""
        return st.session_state.get("user_role")
    
    @staticmethod
    def has_permission(permission: str) -> bool:
        """Check if user has specific permission"""
        user = AuthService.get_current_user()
        if not user:
            return False
        
        permissions = user.get("permissions", [])
        return permission in permissions


# Singleton instance
auth_service = AuthService()

File: frontend/src/services/auth.service.ts
import { apiClient } from '@/lib/api-client';

export interface LoginCredentials {
    username: string;
    password: string;
}

export interface RegisterData {
    username: string;
    email: string;
    password: string;
    first_name?: string;
    last_name?: string;
}

export interface AuthResponse {
    user: User;
    access: string;
    refresh: string;
    access_expires_in: number;
    refresh_expires_in: number;
}

export interface User {
    id: number;
    username: string;
    email: string;
    first_name: string;
    last_name: string;
    role: string;
    is_active: boolean;
}

class AuthService {
    async login(credentials: LoginCredentials): Promise<AuthResponse> {
        const response = await apiClient.post<AuthResponse>('/auth/login/', credentials);

        // Store tokens
        if (typeof window !== 'undefined') {
            localStorage.setItem('access_token', response.access);
            localStorage.setItem('refresh_token', response.refresh);
            localStorage.setItem('user', JSON.stringify(response.user));
        }

        return response;
    }

    async register(data: RegisterData): Promise<User> {
        const response = await apiClient.post<User>('/auth/register/', data);
        return response;
    }

    async logout(): Promise<void> {
        try {
            await apiClient.post('/auth/logout/');
        } catch (error) {
            console.error('Logout error:', error);
        } finally {
            if (typeof window !== 'undefined') {
                localStorage.removeItem('access_token');
                localStorage.removeItem('refresh_token');
                localStorage.removeItem('user');
            }
        }
    }

    async getCurrentUser(): Promise<User> {
        return apiClient.get<User>('/auth/me/');
    }

    isAuthenticated(): boolean {
        if (typeof window === 'undefined') return false;
        return !!localStorage.getItem('access_token');
    }

    getAccessToken(): string | null {
        if (typeof window === 'undefined') return null;
        return localStorage.getItem('access_token');
    }

    getStoredUser(): User | null {
        if (typeof window === 'undefined') return null;
        const userStr = localStorage.getItem('user');
        return userStr ? JSON.parse(userStr) : null;
    }
}

export const authService = new AuthService();
export default authService;
